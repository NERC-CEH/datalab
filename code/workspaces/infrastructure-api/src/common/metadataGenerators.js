import { findKey } from 'lodash';

const stackSecretMetadata = () => ({ annotations: { 'datalab/type': 'stack-secret' } });

const checkResourceHasMetadataFromGenerator = (resource, generator) => {
  const requiredMetadata = generator();
  return findKey(resource, requiredMetadata) === 'metadata';
};

export default {
  stackSecretMetadata,
  checkResourceHasMetadataFromGenerator,
};
