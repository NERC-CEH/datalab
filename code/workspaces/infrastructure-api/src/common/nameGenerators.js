const generateDeploymentName = (name, type) => `${type}-${name}`;
const pvcName = volumeName => `${volumeName}-claim`;

// project namespaces
const projectNamespace = projectKey => projectKey;
const projectComputeNamespace = projectKey => `${projectKey}-compute`;

// Compute submission
const computeSubmissionServiceAccount = projectKey => `${projectKey}-compute-submission-account`;
const computeSubmissionServiceAccountRoleBinding = projectKey => `${computeSubmissionServiceAccount(projectKey)}-role-binding`;
const computeSubmissionClusterRole = () => 'compute-submission-role';

// Spark
const pySparkConfigMap = deploymentName => `${deploymentName}-pyspark-config`;
const sparkDriverHeadlessService = deploymentServiceName => `${deploymentServiceName}-spark-driver-headless-service`;
const sparkJob = deploymentName => `${deploymentName}-spark-job`;

export default {
  deploymentName: generateDeploymentName,
  computeSubmissionClusterRole,
  computeSubmissionServiceAccount,
  computeSubmissionServiceAccountRoleBinding,
  projectNamespace,
  projectComputeNamespace,
  pvcName,
  pySparkConfigMap,
  sparkDriverHeadlessService,
  sparkJob,
};
