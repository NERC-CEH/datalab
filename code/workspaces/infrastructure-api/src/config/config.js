import convict from 'convict';

const config = convict({
  logLevel: {
    doc: 'The logging level',
    format: 'String',
    default: 'info',
    env: 'LOG_LEVEL',
  },
  datalabDomain: {
    doc: 'The domain the datalabs instance is running on',
    format: 'String',
    default: 'datalabs.localhost',
    env: 'DATALAB_DOMAIN',
  },
  apiPort: {
    doc: 'The port for the API service',
    format: 'port',
    default: 8003,
    env: 'INFRASTRUCTURE_API_PORT',
  },
  vaultApi: {
    doc: 'The endpoint for Vault',
    format: 'url',
    default: 'http://localhost:8200',
    env: 'VAULT_API',
  },
  vaultAppRole: {
    doc: 'The Vault app role for the API',
    format: 'String',
    default: 'undefinedrole',
    env: 'VAULT_APP_ROLE',
  },
  kubernetesApi: {
    doc: 'The endpoint for Kubernetes',
    format: 'url',
    default: 'http://localhost:8001',
    env: 'KUBERNETES_API',
  },
  authSigninUrl: {
    doc: 'The sign in URL',
    format: 'url',
    default: 'https://testlab.test-datalabs.nerc.ac.uk',
    env: 'AUTH_SIGNIN_URL',
  },
  authorisationService: {
    doc: 'The endpoint for the authorisation service',
    format: 'url',
    default: 'http://localhost:9000',
    env: 'AUTHORISATION_SERVICE',
  },
  authorisationServiceStub: {
    doc: 'If the authorisation service is in stubbed mode',
    format: 'Boolean',
    default: false,
    env: 'AUTHORISATION_SERVICE_STUB',
  },
  authorisationAudience: {
    doc: 'Expected audience for the authorisation service',
    format: 'url',
    default: 'https://api.datalabs.nerc.ac.uk/',
    env: 'AUTHORISATION_AUDIENCE',
  },
  authorisationIssuer: {
    doc: 'Issuer for the authorisation service to manage',
    format: 'url',
    default: 'https://authorisation.datalabs.nerc.ac.uk/',
    env: 'AUTHORISATION_ISSUER',
  },
  databaseHost: {
    doc: 'The database hostname',
    format: 'String',
    default: 'localhost',
    env: 'DATABASE_HOST',
  },
  databaseUser: {
    doc: 'The user to authenticate against the database',
    format: 'String',
    default: 'datalabs-root-u',
    env: 'DATABASE_USER',
  },
  databasePassword: {
    doc: 'The password for the user being authenticated against the database',
    format: 'String',
    default: 'datalabs-root-p',
    env: 'DATABASE_PASSWORD',
  },
  statusCheckInterval: {
    doc: 'Internal (ms) for querying kubernetes pod status',
    format: 'int',
    default: 60000,
    env: 'STATUS_CHECK_INTERVAL',
  },
  maxBodySize: {
    doc: 'Set value for the maximum upload size permitted',
    format: 'String',
    default: '50g',
    env: 'MAX_BODY_SIZE',
  },
  containerInfoPath: {
    doc: 'The path to yaml file containing container information',
    format: 'String',
    default: './resources/container-info.yml',
    env: 'CONTAINER_INFO_PATH',
  },
});

export default config;
