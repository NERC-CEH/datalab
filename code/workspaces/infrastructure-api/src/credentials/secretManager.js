import vault from './vault';
import tokenGenerator from './tokenGenerator';
import metadataGenerators from '../common/metadataGenerators';
import k8sSecretApi from '../kubernetes/secretApi';
import nameGenerators from '../common/nameGenerators';

function createNewJupyterCredentials() {
  return {
    token: tokenGenerator.generateUUID(),
  };
}

function createNewMinioCredentials() {
  return {
    access_key: tokenGenerator.generateUUID(),
    secret_key: tokenGenerator.generateUUID(),
  };
}

function createNewUserCredentials() {
  return {
    username: 'datalab',
    password: tokenGenerator.generateUUID(),
  };
}

function storeCredentialsInVault(projectKey, id, secret) {
  return vault.ensureSecret(`${projectKey}/stacks/${id}`, secret);
}

function createStackCredentialSecret(stackName, stackType, projectKey, credentials) {
  const secretName = nameGenerators.stackCredentialSecret(stackName, stackType);
  const additionalMetadataForSecret = metadataGenerators.stackSecretMetadata();
  return k8sSecretApi.createOrUpdateSecret(secretName, projectKey, credentials, additionalMetadataForSecret);
}

function deleteStackCredentialSecret(stackName, stackType, projectKey) {
  const secretName = nameGenerators.stackCredentialSecret(stackName, stackType);
  return k8sSecretApi.deleteSecret(secretName, projectKey);
}

function deleteSecret(projectKey, id) {
  return vault.deleteSecret(`${projectKey}/stacks/${id}`);
}

function storeMinioCredentialsInVault(projectKey, id, strategy) {
  return vault.ensureSecret(`${projectKey}/storage/${id}`, strategy);
}

function deleteMinioCredentials(projectKey, id) {
  return vault.deleteSecret(`${projectKey}/storage/${id}`);
}

export default {
  createNewJupyterCredentials,
  createNewMinioCredentials,
  createNewUserCredentials,
  storeCredentialsInVault,
  deleteSecret,
  storeMinioCredentialsInVault,
  deleteMinioCredentials,
  createStackCredentialSecret,
  deleteStackCredentialSecret,
};
