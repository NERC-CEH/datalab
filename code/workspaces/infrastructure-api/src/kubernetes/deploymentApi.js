import axios from 'axios';
import logger from '../config/logger';
import config from '../config/config';
import { handleCreateError, handleDeleteError } from './core';

const API_BASE = config.get('kubernetesApi');

const getDeploymentUrl = (namespace, name) => {
  const nameComponent = name ? `/${name}` : '';
  return `${API_BASE}/apis/apps/v1/namespaces/${namespace}/deployments${nameComponent}`;
};

const getDeploymentScaleUrl = (namespace, name) => `${getDeploymentUrl(namespace, name)}/scale`;

const YAML_CONTENT_HEADER = { headers: { 'Content-Type': 'application/yaml' } };

const PATCH_CONTENT_HEADER = { headers: { 'Content-Type': 'application/merge-patch+json' } };

function createOrUpdateDeployment(name, namespace, manifest) {
  return getDeployment(name, namespace)
    .then(createOrReplace(name, namespace, manifest));
}

const createOrReplace = (name, namespace, manifest) => (existingDeployment) => {
  if (existingDeployment) {
    return updateDeployment(name, namespace, manifest);
  }

  return createDeployment(name, namespace, manifest);
};

function getDeployment(name, namespace) {
  return axios.get(getDeploymentUrl(namespace, name))
    .then(response => response.data)
    .catch(() => undefined);
}

function createDeployment(name, namespace, manifest) {
  logger.info('Creating deployment: %s in namespace: %s', name, namespace);
  return axios.post(getDeploymentUrl(namespace), manifest, YAML_CONTENT_HEADER)
    .catch(handleCreateError('deployment', name));
}

function updateDeployment(name, namespace, manifest) {
  logger.info('Updating deployment: %s in namespace: %s', name, namespace);
  return axios.put(getDeploymentUrl(namespace, name), manifest, YAML_CONTENT_HEADER)
    .catch(handleCreateError('deployment', name));
}

function deleteDeployment(name, namespace) {
  logger.info('Deleting deployment: %s in namespace: %s', name, namespace);
  const deleteOptions = {
    kind: 'DeleteOptions',
    apiVersion: 'v1',
    propagationPolicy: 'Foreground',
  };

  return axios.delete(getDeploymentUrl(namespace, name), { data: deleteOptions })
    .then(response => response.data)
    .catch(handleDeleteError('deployment', name));
}

async function getStatusReplicas(name, namespace) {
  const response = await axios.get(getDeploymentScaleUrl(namespace, name));
  const replicas = response.data.status.replicas || 0;
  return replicas;
}

async function getSpecReplicas(name, namespace) {
  const response = await axios.get(getDeploymentScaleUrl(namespace, name));
  const replicas = response.data.spec.replicas || 0;
  return replicas;
}

async function setSpecReplicas(name, namespace, specReplicas) {
  const response = await axios.patch(
    getDeploymentScaleUrl(namespace, name),
    { spec: { replicas: specReplicas } },
    PATCH_CONTENT_HEADER,
  );

  // Wait for a total of upto approx. 10s for desired number of replicas.
  // Wait 100ms between checks.
  // Waiting helps to prevents http 409 (conflict) errors and ensures the pods are all stopped before restarting.
  const totalWaitMs = 10 * 1000;
  const checkWaitMs = 100;
  for (let i = 0; i < totalWaitMs / checkWaitMs; i += 1) {
    // eslint-disable-next-line no-await-in-loop
    if (specReplicas === await getStatusReplicas(name, namespace)) {
      // specified number of replicas matches status number of replicas
      return response;
    }
    // eslint-disable-next-line no-await-in-loop
    await new Promise(resolve => setTimeout(resolve, checkWaitMs));
  }

  return response;
}

async function restartDeployment(name, namespace) {
  logger.info('Restarting deployment: %s in namespace: %s', name, namespace);

  const initialSpecReplicas = await getSpecReplicas(name, namespace);
  if (initialSpecReplicas > 0) {
    await setSpecReplicas(name, namespace, 0);
  }
  const newSpecReplicas = initialSpecReplicas || 1;
  const response = await setSpecReplicas(name, namespace, newSpecReplicas);
  return response;
}

export default { getDeployment, createDeployment, deleteDeployment, updateDeployment, createOrUpdateDeployment, restartDeployment };
