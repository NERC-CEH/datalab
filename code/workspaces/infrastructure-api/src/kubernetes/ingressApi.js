import axios from 'axios';
import logger from '../config/logger';
import config from '../config/config';
import { handleCreateError, handleDeleteError, handlePatchError } from './core';

const API_BASE = config.get('kubernetesApi');

const getIngressUrl = (namespace, name) => {
  const nameComponent = name ? `/${name}` : '';
  return `${API_BASE}/apis/networking.k8s.io/v1beta1/namespaces/${namespace}/ingresses${nameComponent}`;
};

const YAML_CONTENT_HEADER = { headers: { 'Content-Type': 'application/yaml' } };

const PATCH_CONTENT_HEADER = { headers: { 'Content-Type': 'application/merge-patch+json' } };

function createOrUpdateIngress(name, namespace, manifest) {
  return getIngress(name, namespace)
    .then(createOrReplace(name, namespace, manifest));
}

const createOrReplace = (name, namespace, manifest) => (existingIngress) => {
  if (existingIngress) {
    return updateIngress(name, namespace, manifest);
  }

  return createIngress(name, namespace, manifest);
};

function getIngress(name, namespace) {
  return axios.get(getIngressUrl(namespace, name))
    .then(response => response.data)
    .catch(() => undefined);
}

function createIngress(name, namespace, manifest) {
  logger.info('Creating ingress: %s in namespace %s', name, namespace);
  return axios.post(getIngressUrl(namespace), manifest, YAML_CONTENT_HEADER)
    .catch(handleCreateError('ingress', name));
}

function updateIngress(name, namespace, manifest) {
  logger.info('Updating ingress: %s in namespace %s', name, namespace);
  return axios.put(getIngressUrl(namespace, name), manifest, YAML_CONTENT_HEADER)
    .catch(handleCreateError('ingress', name));
}

async function patchIngress(name, namespace, patchBody) {
  logger.info(`Patching ingress ${name} in namespace ${namespace}`);
  try {
    return await axios.patch(
      getIngressUrl(namespace, name),
      patchBody,
      PATCH_CONTENT_HEADER,
    );
  } catch (error) {
    return handlePatchError('ingress', name)(error);
  }
}

function deleteIngress(name, namespace) {
  logger.info('Deleting ingress: %s in namespace %s', name, namespace);
  return axios.delete(getIngressUrl(namespace, name))
    .then(response => response.data)
    .catch(handleDeleteError('ingress', name));
}

export default { getIngress, createIngress, deleteIngress, updateIngress, patchIngress, createOrUpdateIngress };
