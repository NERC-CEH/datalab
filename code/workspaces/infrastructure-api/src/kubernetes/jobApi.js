import axios from 'axios';
import logger from '../config/logger';
import config from '../config/config';
import { handleCreateError } from './core';

const API_BASE = config.get('kubernetesApi');

const getJobUrl = (namespace, name) => {
  const nameComponent = name ? `/${name}` : '';
  return `${API_BASE}/apis/batch/v1/namespaces/${namespace}/jobs${nameComponent}`;
};

const YAML_CONTENT_HEADER = { headers: { 'Content-Type': 'application/yaml' } };

const createJob = async (name, namespace, manifest) => {
  logger.info('Creating job: %s in namespace: %s', name, namespace);

  try {
    return await axios.post(getJobUrl(namespace), manifest, YAML_CONTENT_HEADER);
  } catch (e) {
    return handleCreateError('job', name)(e);
  }
};

export default {
  createJob,
};
