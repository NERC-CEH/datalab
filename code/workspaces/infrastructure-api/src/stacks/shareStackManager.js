import { stackTypes } from 'common';
import { NOTEBOOK_CATEGORY, SITE_CATEGORY } from 'common/src/config/images';
import config from '../config/config';
import secretManager from '../credentials/secretManager';
import { createKubectlJob } from '../kubernetes/jobGenerator';
import jobApi from '../kubernetes/jobApi';
import ingressApi from '../kubernetes/ingressApi';
import deploymentApi from '../kubernetes/deploymentApi';
import nameGenerators from '../common/nameGenerators';
import { visibility } from '../models/stackEnums';
import zeppelin from './zeppelinStack';

const { deploymentName } = nameGenerators;
const { JUPYTER, JUPYTERLAB, ZEPPELIN } = stackTypes;

const mountPath = '/mnt/persistentfs';

const getJupyterCookiePath = jupyterDirectory => `${jupyterDirectory}/runtime/jupyter_cookie_secret`;

const jupyterDirectoryEnvName = 'JUPYTER_DATA_DIR';

// Find the location of the .jupyter directory based on the environment variables in the deployment.
const getJupyterDirectory = (deployment, deploymentData) => {
  const { containers } = deploymentData.spec.template.spec;
  const defaultDirectory = `${mountPath}/notebooks/${deployment}/.jupyter`;

  const container = containers.find(c => c.name === deployment);
  if (!container) {
    return defaultDirectory;
  }

  const { env } = container;

  const envVar = env.find(e => e.name === jupyterDirectoryEnvName);
  if (!envVar || !envVar.value) {
    return defaultDirectory;
  }

  return envVar.value.replace('/data', mountPath);
};

const getIngressPatch = authValue => ({
  metadata: {
    annotations: {
      'nginx.ingress.kubernetes.io/auth-url': authValue,
    },
  },
});

export const makeJupyterPrivate = async (name, type, projectKey, volumeMount) => {
  // Changing a Jupyter notebook to private requires a token refresh, and to remove a file from the notebook's storage.
  const credentials = secretManager.createNewJupyterCredentials();
  await secretManager.createStackCredentialSecret(name, type, projectKey, credentials);

  const deployment = deploymentName(name, type);

  const deploymentData = await deploymentApi.getDeployment(deployment, projectKey);
  const jupyterDirectory = getJupyterDirectory(deployment, deploymentData);

  // Trigger a job to remove a cookie file from the volume and restart the notebook pod
  const runCommand = `rm -f ${getJupyterCookiePath(jupyterDirectory)}`;
  const kubectlCommand = `kubectl rollout restart deployment/${deployment}`;

  const manifest = await createKubectlJob({ name, runCommand, kubectlCommand, volumeMount, mountPath });
  await jobApi.createJob(name, projectKey, manifest);
};

export const makeZeppelinPrivate = async (name, type, projectKey) => {
  const credentials = secretManager.createNewUserCredentials();

  const shiroIni = await zeppelin.generateNewShiroIni(credentials);
  await secretManager.createStackCredentialSecret(name, type, projectKey, { ...shiroIni, ...credentials });

  const deployment = deploymentName(name, type);
  await deploymentApi.restartDeployment(deployment, projectKey);
};

export const handleSharedChange = async (params, existing, newSharedStatus) => {
  const { category, shared, type, volumeMount, visible } = existing;

  const oldSharedStatus = shared || visible;

  if (oldSharedStatus === visibility.PRIVATE || oldSharedStatus === newSharedStatus) {
    // Going from private to another status (or staying at the same status) requires no extra changes.
    return;
  }

  const { projectKey, name } = params;

  if (category === SITE_CATEGORY) {
    const ingressName = deploymentName(name, type);
    if (newSharedStatus === visibility.PUBLIC) {
      // When making a site public, remove an ingress annotation
      await ingressApi.patchIngress(ingressName, projectKey, getIngressPatch(null));
    }

    if (oldSharedStatus === visibility.PUBLIC) {
      // When making a site non-public, add an ingress annotation
      const authServiceUrlRoot = config.get('authorisationServiceForIngress') || config.get('authorisationService');
      await ingressApi.patchIngress(ingressName, projectKey, getIngressPatch(`${authServiceUrlRoot}/auth`));
    }

    // No backend changes needed for other status changes
    return;
  }

  if (category === NOTEBOOK_CATEGORY && newSharedStatus === visibility.PRIVATE) {
    if (type === JUPYTER || type === JUPYTERLAB) {
      await makeJupyterPrivate(name, type, projectKey, volumeMount);
    }

    if (type === ZEPPELIN) {
      await makeZeppelinPrivate(name, type, projectKey);
    }

    // Nothing to do for RStudio, we just have to let the token expire.
  }
};

export default {
  handleSharedChange,
};

