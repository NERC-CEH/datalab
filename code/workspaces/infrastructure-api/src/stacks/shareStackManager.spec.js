import { stackTypes } from 'common';
import { NOTEBOOK_CATEGORY, SITE_CATEGORY } from 'common/src/config/images';
import { getCurlCommand, handleSharedChange } from './shareStackManager';
import config from '../config/config';
import secretManager from '../credentials/secretManager';
import jobApi from '../kubernetes/jobApi';
import ingressApi from '../kubernetes/ingressApi';
import deploymentApi from '../kubernetes/deploymentApi';
import zeppelin from './zeppelinStack';

jest.mock('../credentials/secretManager');
jest.mock('../kubernetes/jobApi');
jest.mock('../kubernetes/ingressApi');
jest.mock('../kubernetes/deploymentApi');
jest.mock('./zeppelinStack');

const projectKey = 'project';
const name = 'stackname';

const volumeMount = 'volume';

const { JUPYTERLAB, ZEPPELIN, RSTUDIO, RSHINY } = stackTypes;

const getParams = () => ({
  projectKey,
  name,
});
const getExisting = () => ({
  shared: 'project',
  visible: 'project',
  volumeMount,
});
const userToken = 'Bearer token';

// Function to help 'expect' numbers of calls for certain functions (so we don't need to replicate this several times).
const expectCalls = ({
  createJob = 0,
  patchIngress = 0,
  createStackCredentialSecret = 0,
  getDeployment = 0,
  restartDeployment = 0,
  generateNewShiroIni = 0,
}) => {
  expect(jobApi.createJob).toHaveBeenCalledTimes(createJob);
  expect(ingressApi.patchIngress).toHaveBeenCalledTimes(patchIngress);
  expect(secretManager.createStackCredentialSecret).toHaveBeenCalledTimes(createStackCredentialSecret);
  expect(deploymentApi.getDeployment).toHaveBeenCalledTimes(getDeployment);
  expect(deploymentApi.restartDeployment).toHaveBeenCalledTimes(restartDeployment);
  expect(zeppelin.generateNewShiroIni).toHaveBeenCalledTimes(generateNewShiroIni);
};

describe('handleSharedChange', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('does nothing if the old status is private', async () => {
    const existing = {
      ...getExisting(),
      shared: 'private',
    };
    await handleSharedChange(getParams(), existing, 'project', userToken);

    expectCalls({});
  });

  it('does nothing if the new status is the same as the old status', async () => {
    await handleSharedChange(getParams(), getExisting(), 'project');

    expectCalls({});
  });

  it('does nothing if nothing is to be done for the site category', async () => {
    const existing = {
      ...getExisting(),
      category: 'other',
    };
    await handleSharedChange(getParams(), existing, 'public', userToken);

    expectCalls({});
  });

  it('does nothing for RStudio notebooks', async () => {
    const existing = {
      ...getExisting(),
      category: NOTEBOOK_CATEGORY,
      type: RSTUDIO,
    };
    await handleSharedChange(getParams(), existing, 'private', userToken);

    expectCalls({});
  });

  it('does nothing for Sites that are not changing to or from public', async () => {
    const existing = {
      ...getExisting(),
      category: SITE_CATEGORY,
      type: RSHINY,
    };
    await handleSharedChange(getParams(), existing, 'private', userToken);

    expectCalls({});
  });

  it('handles Sites changing to public', async () => {
    const existing = {
      ...getExisting(),
      category: SITE_CATEGORY,
      type: RSHINY,
    };
    const expectedPatch = {
      metadata: {
        annotations: {
          'nginx.ingress.kubernetes.io/auth-url': null,
        },
      },
    };

    await handleSharedChange(getParams(), existing, 'public', userToken);

    expectCalls({ patchIngress: 1 });
    expect(ingressApi.patchIngress).toHaveBeenCalledWith('rshiny-stackname', projectKey, expectedPatch);
  });

  it('handles Sites changing from public', async () => {
    const existing = {
      ...getExisting(),
      category: SITE_CATEGORY,
      type: RSHINY,
      shared: 'public',
      visible: 'public',
    };
    jest.spyOn(config, 'get').mockReturnValue('configUrl');

    const expectedPatch = {
      metadata: {
        annotations: {
          'nginx.ingress.kubernetes.io/auth-url': 'configUrl/auth',
        },
      },
    };

    await handleSharedChange(getParams(), existing, 'project', userToken);

    expectCalls({ patchIngress: 1 });
    expect(ingressApi.patchIngress).toHaveBeenCalledWith('rshiny-stackname', projectKey, expectedPatch);
  });

  it('handles Jupyter notebooks changing to private', async () => {
    const existing = {
      ...getExisting(),
      category: NOTEBOOK_CATEGORY,
      type: JUPYTERLAB,
    };
    jest.spyOn(Date, 'now').mockReturnValue(1609459200);

    secretManager.createNewJupyterCredentials = jest.fn(() => ({ token: 'token' }));
    deploymentApi.getDeployment.mockResolvedValueOnce({
      spec: {
        template: {
          spec: {
            containers: [
              {
                name: 'jupyterlab-stackname',
                env: [
                  {
                    name: 'JUPYTER_DATA_DIR',
                    value: '/data/notebooks/jupyterlab-stackname/.jupyter',
                  },
                ],
              },
            ],
          },
        },
      },
    });

    await handleSharedChange(getParams(), existing, 'private', userToken);

    expectCalls({ createJob: 1, createStackCredentialSecret: 1, getDeployment: 1 });
    expect(secretManager.createStackCredentialSecret).toHaveBeenCalledWith(name, JUPYTERLAB, projectKey, { token: 'token' });
    expect(deploymentApi.getDeployment).toHaveBeenCalledWith('jupyterlab-stackname', projectKey);
    expect(jobApi.createJob).toHaveBeenCalledWith(name, projectKey, expect.any(String));
    expect(jobApi.createJob.mock.calls[0][2]).toMatchSnapshot();
  });

  it('handles Zeppelin notebooks changing to private', async () => {
    const existing = {
      ...getExisting(),
      category: NOTEBOOK_CATEGORY,
      type: ZEPPELIN,
    };

    const expectedCredentials = {
      username: 'datalab',
      password: 'password',
    };
    secretManager.createNewUserCredentials = jest.fn(() => expectedCredentials);
    zeppelin.generateNewShiroIni.mockResolvedValueOnce({ 'shiro.ini': 'shiro' });

    const expectedFullCredentials = {
      'shiro.ini': 'shiro',
      ...expectedCredentials,
    };

    await handleSharedChange(getParams(), existing, 'private', userToken);

    expectCalls({ createStackCredentialSecret: 1, restartDeployment: 1, generateNewShiroIni: 1 });
    expect(zeppelin.generateNewShiroIni).toHaveBeenCalledWith(expectedCredentials);
    expect(secretManager.createStackCredentialSecret).toHaveBeenCalledWith(name, ZEPPELIN, projectKey, expectedFullCredentials);
    expect(deploymentApi.restartDeployment).toHaveBeenCalledWith('zeppelin-stackname', projectKey);
  });
});

describe('getCurlCommand', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('formats the request correctly', () => {
    const mockConfig = {
      apiPort: 8000,
      deployedNamespace: 'test',
      deployedInCluster: true,
    };
    jest.spyOn(config, 'get').mockReturnValueOnce(mockConfig);

    // eslint-disable-next-line max-len
    const expected = 'curl -X PUT infrastructure-api-service.test.svc.cluster.local:8000/stack/project/restart -H "Authorization: $TOKEN" -H "Content-Type: application/json" -d \'\'{"projectKey":"project","name":"stackname","type":"jupyterlab"}\'\'';

    expect(getCurlCommand('project', 'stackname', 'jupyterlab')).toEqual(expected);
  });

  it('formats the request correctly for running locally', () => {
    const mockConfig = {
      apiPort: 8000,
      deployedNamespace: 'test',
      deployedInCluster: false,
    };
    jest.spyOn(config, 'get').mockReturnValueOnce(mockConfig);

    // eslint-disable-next-line max-len
    const expected = 'curl -X PUT host.docker.internal:8000/stack/project/restart -H "Authorization: $TOKEN" -H "Content-Type: application/json" -d \'\'{"projectKey":"project","name":"stackname","type":"jupyterlab"}\'\'';

    expect(getCurlCommand('project', 'stackname', 'jupyterlab')).toEqual(expected);
  });
});

