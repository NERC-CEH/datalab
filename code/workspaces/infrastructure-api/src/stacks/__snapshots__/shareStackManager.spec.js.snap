// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handleSharedChange handles Jupyter notebooks changing to private 1`] = `
"---
apiVersion: batch/v1
kind: Job
metadata:
  name: job-stackname
spec:
  ttlSecondsAfterFinished: 0
  template:
    spec:
      securityContext:
        runAsUser: 1000
      containers:
      - name: job-stackname
        image: busybox
        imagePullPolicy: IfNotPresent
        command: [\\"sh\\"]
        args: [\\"-c\\", \\"rm -f /mnt/persistentfs/notebooks/jupyterlab-stackname/.jupyter/runtime/jupyter_cookie_secret\\"]
        volumeMounts:
          - mountPath: /mnt/persistentfs
            name: persistentfsvol
      - name: curl-job
        image: curlimages/curl
        imagePullPolicy: IfNotPresent
        command: [\\"sh\\"]
        # Command in single quotes as there are double quotes in the command itself
        args: [\\"-c\\", 'curl -X PUT host.docker.internal:undefined/stack/project/restart -H \\"Authorization: $TOKEN\\" -H \\"Content-Type: application/json\\" -d ''{\\"projectKey\\":\\"project\\",\\"name\\":\\"stackname\\",\\"type\\":\\"jupyterlab\\"}''']
        env:
        - name: TOKEN
          value: \\"Bearer token\\"
      volumes:
        - name: persistentfsvol
          persistentVolumeClaim:
            claimName: volume-claim
      restartPolicy: Never
"
`;
